server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,962 @@
+    // Any other error report failure.
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+// Initiate copy of source part object to the destination
+void S3PutMultipartCopyAction::initiate_part_copy() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  max_parallel_copy = MAX_PARALLEL_COPY;
+  if (additional_object_metadata->get_number_of_fragments() == 0) {
+    copy_part_object();
+  } else {
+    s3_copy_part_action_state = S3CopyPartActionState::validationFailed;
+    set_s3_error("NotImplemented");
+    send_response_to_s3_client();
+  }
+}
+
+// Copy source part object to destination
+void S3PutMultipartCopyAction::copy_part_object() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  if (!total_data_to_stream) {
+    s3_log(S3_LOG_DEBUG, stripped_request_id, "Source object is empty");
+    next();
+    return;
+  }
+  bool f_success = false;
+  try {
+    object_data_copier.reset(new S3ObjectDataCopier(
+        request, motr_writer, motr_reader_factory, s3_motr_api));
+
+    object_data_copier->copy(
+        additional_object_metadata->get_oid(), total_data_to_stream,
+        additional_object_metadata->get_layout_id(),
+        additional_object_metadata->get_pvid(),
+        std::bind(&S3PutMultipartCopyAction::copy_part_object_cb, this),
+        std::bind(&S3PutMultipartCopyAction::copy_part_object_success, this),
+        std::bind(&S3PutMultipartCopyAction::copy_part_object_failed, this));
+    f_success = true;
+  }
+  catch (const std::exception& ex) {
+    s3_log(S3_LOG_ERROR, stripped_request_id, "%s", ex.what());
+  }
+  catch (...) {
+    s3_log(S3_LOG_ERROR, stripped_request_id, "Non-standard C++ exception");
+  }
+  if (!f_success) {
+    object_data_copier.reset();
+
+    set_s3_error("InternalError");
+    send_response_to_s3_client();
+    return;
+  }
+  if (total_data_to_stream > MINIMUM_ALLOWED_PART_SIZE) {
+    start_response();
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::copy_part_object_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  s3_copy_part_action_state = S3CopyPartActionState::writeComplete;
+  object_data_copier.reset();
+  next();
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::copy_part_object_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  s3_copy_part_action_state = S3CopyPartActionState::writeFailed;
+  set_s3_error(object_data_copier->get_s3_error());
+
+  object_data_copier.reset();
+  send_response_to_s3_client();
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::save_metadata() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  std::string s_md5_got = request->get_header_value("content-md5");
```

Comment:
Check https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html, I feel this header is not there

Reply:
Yes, removed this header.